assign-doi-datacite/src/main/java/no/unit/nva/datacite/clients/DataciteClient.java

```diff
@@ -0,0 +1,161 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A DoiClient implementation towards Registry Agency Datacite.
+ */
+public class DataciteClient implements DoiClient {
+
+    public static final String ERROR_SETTING_DOI_METADATA_TEMPLATE = "Error setting DOI metadata (%s)";
+    public static final String ERROR_SETTING_DOI_URL_TEMPLATE = "Error setting DOI url (%s)";
+    public static final String ERROR_DELETING_DOI_METADATA_TEMPLATE = "Error deleting DOI metadata (%s)";
+    public static final String ERROR_DELETING_DOI_TEMPLATE = "Error deleting DOI (%s)";
+    public static final String ERROR_COMMUNICATION_TEMPLATE = "Error during API communication: (%s)";
+    protected static final String CHARACTER_PARENTHESES_START = "(";
+    protected static final String CHARACTER_PARENTHESES_STOP = ")";
+    protected static final String CHARACTER_WHITESPACE = " ";
+    protected static final char FORWARD_SLASH = '/';
+    private static final Logger logger = LoggerFactory.getLogger(DataciteClient.class);
+    private final DataciteMdsConnectionFactory mdsConnectionFactory;
+    private final DataciteConfigurationFactory configFactory;
+
+    public DataciteClient(DataciteConfigurationFactory configFactory,
+                          DataciteMdsConnectionFactory mdsConnectionFactory) {
+        this.configFactory = configFactory;
+        this.mdsConnectionFactory = mdsConnectionFactory;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Doi createDoi(String customerId, String metadataDataciteXml) throws ClientException {
+        var prefix = configFactory.getConfig(customerId)
+            .map(DataCiteMdsClientConfig::getInstitutionPrefix)
+            .orElseThrow();
+        Doi doi;
+        try {
+            var response =
+                mdsConnectionFactory
+                    .getAuthenticatedConnection(customerId)
+                    .postMetadata(prefix, metadataDataciteXml);
+            if (!isSuccessfulApiResponse(response)) {
+                throw logAndCreateApiException(response.statusCode(), ERROR_SETTING_DOI_METADATA_TEMPLATE);
+            }
+            String createMetadataResponseBody = response.body();
+            doi = extractDoiPrefixAndSuffix(createMetadataResponseBody);
+        } catch (IOException | URISyntaxException | InterruptedException e) {
+            throw logAndCreateClientException("createDoi", e);
+        }
+        return doi;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void updateMetadata(String customerId, Doi doi, String metadataDataciteXml) throws ClientException {
+        try {
+            var response = mdsConnectionFactory
+                .getAuthenticatedConnection(customerId)
+                .postMetadata(doi.toIdentifier(), metadataDataciteXml);
+            if (!isSuccessfulApiResponse(response)) {
+                throw new ClientException(ERROR_SETTING_DOI_METADATA_TEMPLATE
```

Comment:
Can we create more specific exceptions so we can hide away these constants?

Reply:
`throw logAndCreateApiException(response.statusCode(), ERROR_SETTING_DOI_METADATA_TEMPLATE);` was the code supposed to be.

Is this good enough or do we actually want a dedicated exception for each method on the client?